[GEOS-10926] Community Module Proxy-Base-Ext

turingtestfail opened the pull request:
[![GEOS-10926](https://badgen.net/badge/JIRA/GEOS-10926/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-10926)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

link and graphic

dao first pass

started on URLMangler

mangler first pass

first test

more tests

more tests2

integration tests added

added first bit of REST controller

more rest controller

Created rule bean builder and updated dao and controller

added position to the rule

added position sort

started on rest unit tests

rest test get put

rest post test

transformer input not working

added back position and fixed issue with model calling builder

ui tests

added copyright

cleaned up test resource paths

javadoc

web page description improvement

REST API YAML

first pass at documentation

added test interface and updated documentation

updated usage

spotless

more cleanup


<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

aaime commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java:
```diff
@@ -0,0 +1,197 @@
+     * @return the transformed URL
+     * @throws IOException if the URL is malformed
+     */
+    public String transformURL(String urlString, String headersAsProperties) throws IOException {
+        URL url = new URL(urlString);
+        StringBuilder baseURL =
+                new StringBuilder(url.getProtocol() + PROTOCOL_SEPARATOR + url.getHost());
```
Actually... according to Jacoco's code coverage, this method is not exercised.... but I see the WEB UI actually calling it. So I'm guessing there are no tests in the web page for it.

turingtestfail replied:
done

aaime commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java:
```diff
@@ -0,0 +1,171 @@
+import org.geoserver.web.wicket.ParamResourceModel;
+
+/** The configuration page for the {@link WebLinkingUrlMangler}. */
+public class WebLinkingManglingConfigPage extends GeoServerSecuredPage {
```
e.g.:

```suggestion
public class ProxyBaseExtConfigPage extends GeoServerSecuredPage {
```

turingtestfail replied:
done

aaime commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingRulePage.java:
```diff
@@ -0,0 +1,100 @@
+         */
+        public SimpleRulePanel(String panelId, IModel<RuleModel> model) {
+            super(panelId, model);
+            add(new NumberTextField<Integer>("position").setMinimum(1).setRequired(true));
```
When creating a new rule the default is "zero":

![image](https://user-images.githubusercontent.com/325433/233657616-3ada786a-6db9-428a-91f8-9eede5e70a35.png)

Actually not sure why the minimum has to be 1, but either make the minimum zero, or make sure the default value is sensible.

turingtestfail replied:
done

turingtestfail commented:
> Testing it interactively, I cannot make it work:
> 
> ![image](https://user-images.githubusercontent.com/325433/233658354-8aded8a3-9413-4fbe-bf9f-fa3fb6557e65.png)
> 
> Why should I select rules? The testing should default on using them all... But even if I select one, it won't work (should, the URL is a match for the regex in the rule...)

Changed to use all rules.  It wasn't working because the protocol is required.  Changed to report URL issues into the output box.

turingtestfail commented on doc/en/user/source/community/proxy-base-ext/install.rst:
```diff
@@ -0,0 +1,24 @@
+
+#. Restart GeoServer.
+
+If installation was successful, you will see a new Proxy Base Extension entry in the left menu, under "Settings".
```
done

turingtestfail commented on doc/en/user/source/community/proxy-base-ext/usage.rst:
```diff
@@ -0,0 +1,66 @@
+
+This extension allows you to replace URLs in the response of a web service request with a different URL.
+This is useful when you want to use a proxy to redirect your web service requests, but you want to keep the original URL in the response.
+An example of this is when you want to proxy a WMS request to a different server, but you want to keep the original URL in the elements of the GetCapabilities response.
```
done

turingtestfail commented on doc/en/user/source/community/proxy-base-ext/usage.rst:
```diff
@@ -0,0 +1,66 @@
+
+Rules can be managed and tested with simulated headers in the rules management UI. Besides the basic operations like add, remove and update is also possible to activate or deactivate rules. A deactivated rule will be ignored by this module.
+
+.. figure:: images/proxy_base_rule_config.png
```
done

turingtestfail commented on doc/en/user/source/community/proxy-base-ext/usage.rst:
```diff
@@ -0,0 +1,66 @@
+.. figure:: images/proxy_base_rule_config.png
+   :align: center
+
+   *Rules management UI*
```
done

turingtestfail commented on src/community/proxy-base-ext/pom.xml:
```diff
@@ -0,0 +1,54 @@
+
+  <groupId>org.geoserver.community</groupId>
+  <artifactId>gs-proxy-base-ext</artifactId>
+  <name>Configurable altering of GeoServer provided web links</name>
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java:
```diff
@@ -0,0 +1,197 @@
+        } catch (MalformedURLException e) {
+            LOGGER.log(
+                    Level.ALL,
+                    "Web Linking URL Mangler: The transformer, after header template replacement: "
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java:
```diff
@@ -0,0 +1,197 @@
+
+    private Map<String, String> collectHeaders(
+            List<String> requiredHeaders, String headersAsProperties) throws IOException {
+        Properties properties = parsePropertiesString(headersAsProperties);
```
Both work, documentation changed to reflect properties format

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleConverter.java:
```diff
@@ -0,0 +1,76 @@
+import org.geoserver.proxybase.ext.LinkingManglerRuleBuilder;
+
+/** Converts a {@link LinkingManglerRule} to and from XML. */
+public class LinkingManglerRuleConverter implements Converter {
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleDAO.java:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return a temporary path for the {@link LinkingManglerRule} file
+     */
+    public static String getTmpLinkingManglerRulesPathPath() {
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/rest/LinkingManglerRulesController.java:
```diff
@@ -0,0 +1,179 @@
+            MediaTypeExtensions.TEXT_JSON_VALUE
+        })
+/** Controller for the {@link LinkingManglerRule}s. */
+public class LinkingManglerRulesController extends RestBaseController {
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RuleModel.java:
```diff
@@ -0,0 +1,110 @@
+import org.geoserver.proxybase.ext.config.LinkingManglerRule;
+
+/** A model for the {@link LinkingManglerRule}. */
+public class RuleModel implements Serializable {
```
removed

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RulesModel.java:
```diff
@@ -0,0 +1,68 @@
+import org.geoserver.web.wicket.GeoServerDataProvider;
+
+/** A model for a collection of the {@link LinkingManglerRule}. */
+public class RulesModel extends GeoServerDataProvider<RuleModel> {
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java:
```diff
@@ -0,0 +1,171 @@
+                                return null;
+                            }
+                        });
+        rulesPanel.setOutputMarkupId(true);
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java:
```diff
@@ -0,0 +1,171 @@
+        return header;
+    }
+
+    private class EditButtonPanel extends Panel {
```
done

turingtestfail commented on src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java:
```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    private class ActivateButtonPanel extends Panel {
```
Checkbox doesn't work as a component, needs to wrapped in a panel.

turingtestfail commented on src/community/proxy-base-ext/src/test/java/org/geoserver/proxybase/ext/WebLinkingUrlManglerRestTest.java:
```diff
@@ -0,0 +1,149 @@
+/** Tests the REST API for the web linking mangler */
+public class WebLinkingUrlManglerRestTest extends WebLinkingUrlManglerTestSupport {
+    @Test
+    public void getGetRulesXML() throws Exception {
```
Validated that auth is required with manual testing, doesn't seem to be required with MockHTTP/GeoServerSystemTestSupport
